### PR TITLE
feat: Pack Hub Structs

### PR DESCRIPTION
**Pull request by avniculae**
Packs structs used by the Hub
- uses uint128 instead of uin256 for stored assets/shares amounts
- uses uint40 for stored timestamps
- uses uint56 for caps (caps are not scaled by the number of asset's decimals anymore)
- uses uint16 for liquidityFee
- inlines AssetConfig & SpokeConfig

### CODE REVIEW

File: src/contracts/Hub.sol [hunk 1]
```diff
@@ -268,10 +270,10 @@ contract Hub is IHub, AccessManaged {
     _validateReportDeficit(asset, spoke, drawnAmount, premiumAmount);
 
     _applyPremiumDelta(asset, spoke, premiumDelta, premiumAmount);
-    uint256 drawnShares = previewRestoreByAssets(assetId, drawnAmount);
+    uint128 drawnShares = previewRestoreByAssets(assetId, drawnAmount).toUint128();
     asset.drawnShares -= drawnShares;
     spoke.drawnShares -= drawnShares;
-    uint256 totalDeficitAmount = drawnAmount + premiumAmount;
+    uint128 totalDeficitAmount = (drawnAmount + premiumAmount).toUint128();
```
Comment (DhairyaSethi) [inline src/contracts/Hub.sol]: could cast later here as well (same as restore)

File: src/interfaces/IBasicInterestRateStrategy.sol [hunk 2]
```diff
@@ -20,5 +20,5 @@ interface IBasicInterestRateStrategy {
     uint256 liquidity,
     uint256 drawn,
     uint256 premium
-  ) external view returns (uint256);
+  ) external view returns (uint256 variableBorrowRate);
```
Comment (DhairyaSethi) [inline src/interfaces/IBasicInterestRateStrategy.sol]: just borrow rate is ok, we have no distinction bw stable & variable rate

File: src/libraries/types/DataTypes.sol [hunk 3]
```diff
@@ -5,41 +5,40 @@ import {IHub} from 'src/interfaces/IHub.sol';
...
-    uint256 drawnShares;
-    uint256 premiumShares;
-    uint256 premiumOffset; // todo make signed
-    uint256 realizedPremium;
+    uint128 addedShares;
+    uint128 drawnShares;
+    uint128 premiumShares;
+    uint128 premiumOffset;
+    uint128 realizedPremium;
     DataTypes.SpokeConfig config;
```
Comment (DhairyaSethi) [inline src/libraries/types/DataTypes.sol]: we'd save a cold sload if we were to store this config together in bytes16 or smthing, and have a helper library which extracts this struct OR just move the three vars here directly. thoughts?

File: src/libraries/types/DataTypes.sol [hunk 4]
```diff
@@ -5,41 +5,40 @@ import {IHub} from 'src/interfaces/IHub.sol';
...
+    uint40 lastUpdateTimestamp;
+    uint128 addedShares;
+    uint128 liquidity;
+    uint128 drawnShares;
+    uint128 premiumShares;
+    uint128 premiumOffset;
+    uint128 realizedPremium;
+    uint128 drawnIndex;
+    uint128 drawnRate;
+    uint128 deficit;
```
Comment (DhairyaSethi) [inline src/libraries/types/DataTypes.sol]: pls move this up with liquidity, they are almost always fetched together, similar idea on others

File: src/libraries/types/DataTypes.sol [hunk 5]
```diff
@@ -5,41 +5,40 @@ import {IHub} from 'src/interfaces/IHub.sol';
...
     DataTypes.AssetConfig config;
   }
 
   struct SpokeConfig {
     bool active;
-    uint256 addCap;
-    uint256 drawCap;
+    uint64 addCap;
```
Comment (DhairyaSethi) [inline src/libraries/types/DataTypes.sol]: v3 uses 35 bits, maybe we can make do with 32 but nbd can keep as is bc there's no additional saving since rn this takes a new slot, if we extract it out of the struct even then theres plenty of space in 128 bits & we have no vars after this

File: tests/Base.t.sol [hunk 6]
```diff
@@ -358,8 +358,8 @@ abstract contract Base is Test {
   function configureTokenList() internal {
     DataTypes.SpokeConfig memory spokeConfig = DataTypes.SpokeConfig({
       active: true,
-      addCap: UINT256_MAX,
-      drawCap: UINT256_MAX
+      addCap: type(uint64).max,
```
Comment (DhairyaSethi) [inline tests/Base.t.sol]: maybe we extract this sentinel max cap value out in a constant, but nbd

File: tests/unit/Hub/Hub.Add.t.sol [hunk 7]
```diff
@@ -76,73 +123,38 @@ contract HubAddTest is HubBase {
...
       caller: address(spoke1),
       amount: addedAmount,
       user: alice
     });
 
     // set add cap to amount of assets added * 2 - 1, given
     // that the same asset amount is provided again below
-    uint256 newAddCap = 2 * addedAmount - 1;
-    _updateAddCap(daiAssetId, address(spoke1), newAddCap);
+    uint64 newAddCap = uint64(2 * addedAmount - 1);
```
Comment (DhairyaSethi) [inline tests/unit/Hub/Hub.Add.t.sol]: on quick look, maybe this value overflows 64 bits but not sure so we should safe cast

File: src/contracts/Hub.sol (hunk 1 above)
Comment (avniculae) [reply src/contracts/Hub.sol]: [72538e1](https://github.com/aave/aave-v4/pull/591/commits/72538e1e3ef8b656df18cadbf91d6a8bbafa7bc9)

File: src/interfaces/IBasicInterestRateStrategy.sol (hunk 2 above)
Comment (avniculae) [reply src/interfaces/IBasicInterestRateStrategy.sol]: [72538e1](https://github.com/aave/aave-v4/pull/591/commits/72538e1e3ef8b656df18cadbf91d6a8bbafa7bc9)

File: src/contracts/Hub.sol (hunk 1 above)
Comment (yan-man) [reply src/contracts/Hub.sol]: is there a benefit to casting later rather than init totalDeficitAmount as uint128 directly

File: tests/unit/Hub/Hub.Add.t.sol [hunk 8]
```diff
@@ -6,6 +6,46 @@ import 'tests/unit/Hub/HubBase.t.sol';
 contract HubAddTest is HubBase {
```
Comment (DhairyaSethi) [inline tests/unit/Hub/Hub.Add.t.sol:6]: could you pls add basic tests for all hub methods reverting on safe cast overflow? 
for rate casting we can mock ir strategy return value, idt we can clearly test index cast
but main is just input amount values

File: src/libraries/types/DataTypes.sol (hunk 4 above)
Comment (DhairyaSethi) [reply src/libraries/types/DataTypes.sol]: drawnRate & timestamp can be in the same slot (odd no of uint128's)

File: src/contracts/Hub.sol [hunk 9]
```diff
@@ -122,7 +129,9 @@ contract Hub is IHub, AccessManaged {
     require(!_assetToSpokes[assetId].contains(spoke), SpokeAlreadyListed());
 
     _assetToSpokes[assetId].add(spoke);
-    _spokes[assetId][spoke].config = config;
+    _spokes[assetId][spoke].active = config.active;
+    _spokes[assetId][spoke].addCap = config.addCap;
+    _spokes[assetId][spoke].drawCap = config.drawCap;
```
Comment (DhairyaSethi) [inline src/contracts/Hub.sol]: let's move this + emit spoke config updated in an internal method bc it's repeated

File: src/libraries/logic/AssetLogic.sol [hunk 10]
```diff
@@ -98,14 +101,13 @@ library AssetLogic {
...
   function updateDrawnRate(DataTypes.Asset storage asset, uint256 assetId) internal {
-    uint256 newBorrowRate = IBasicInterestRateStrategy(asset.config.irStrategy)
-      .calculateInterestRate({
-        assetId: assetId,
-        liquidity: asset.liquidity,
-        drawn: asset.drawn(),
-        premium: asset.premium()
-      });
-    asset.drawnRate = newBorrowRate;
+    uint256 newBorrowRate = IBasicInterestRateStrategy(asset.irStrategy).calculateInterestRate({
```
Comment (DhairyaSethi) [inline src/libraries/logic/AssetLogic.sol]:
```suggestion
    uint256 newDrawnRate = IBasicInterestRateStrategy(asset.irStrategy).calculateInterestRate({
```

File: tests/gas/Spoke.Operations.gas.t.sol [hunk 11]
```diff
@@ -102,7 +102,21 @@ contract SpokeOperations_Gas_Tests is Base {
...
+    vm.snapshotGasLastCall('Spoke.Operations', 'borrow: 1 asset borrows, 1 user borrows');
+    vm.stopPrank();
+
+    skip(60);
+
+    vm.startPrank(carol);
+    spoke1.supply(_usdxReserveId(spoke1), 1000e6, carol);
+    spoke1.setUsingAsCollateral(_usdxReserveId(spoke1), true, carol);
+    spoke1.borrow(_daiReserveId(spoke1), 200e18, carol);
+    vm.snapshotGasLastCall('Spoke.Operations', 'borrow: 2 asset borrows, 0 user borrows');
```
Comment (DhairyaSethi) [inline tests/gas/Spoke.Operations.gas.t.sol]: i would remove these, from existing suffix after `:` is for existing state, and it's fine if we're not testing the difference between existing asset state bc it's nbd, we can do `borrow`, and `borrow: second action, same reserve` like supply which is enough, wdyt?

File: src/libraries/types/DataTypes.sol [hunk 12]
```diff
@@ -5,41 +5,44 @@ import {IHub} from 'src/interfaces/IHub.sol';
...
-    uint256 drawnShares;
-    uint256 premiumShares;
-    uint256 premiumOffset; // todo make signed
-    uint256 realizedPremium;
-    uint256 drawnIndex;
-    uint256 drawnRate;
-    uint256 lastUpdateTimestamp;
-    uint256 deficit;
-    DataTypes.AssetConfig config;
+    uint128 drawnRate;
```
Comment (DhairyaSethi) [inline src/libraries/types/DataTypes.sol]: let's try to keep `lastUpdateTimestamp` & `drawnRate` in the same slot - it should be better than it being packed with `underlying` & `decimals` bc the number of slots should still remain the same

File: src/libraries/types/DataTypes.sol [hunk 13]
```diff
@@ -5,41 +5,44 @@ import {IHub} from 'src/interfaces/IHub.sol';
...
 
   struct Asset {
+    uint128 liquidity;
+    uint128 deficit;
+    uint128 addedShares;
+    uint128 drawnShares;
+    uint128 premiumShares;
+    uint128 premiumOffset;
+    uint128 drawnIndex;
+    uint128 realizedPremium;
```
Comment (DhairyaSethi) [inline src/libraries/types/DataTypes.sol]: let's move `drawnIndex` with `drawnShares` 
when we add `sweeped`, ie reinvested counter, then it can be packed with `addedShares`
so we'll have liquidity, deficit; sweeped, addedShares; drawnIndex, drawnShares; premiumShares..

File: src/libraries/types/DataTypes.sol [hunk 14]
```diff
@@ -5,41 +5,44 @@ import {IHub} from 'src/interfaces/IHub.sol';
...
-    uint256 realizedPremium;
-    DataTypes.SpokeConfig config;
+    uint128 addedShares;
+    uint128 drawnShares;
+    uint128 premiumShares;
+    uint128 premiumOffset;
+    uint128 realizedPremium;
+    uint64 addCap;
+    uint64 drawCap;
+    bool active;
```
Comment (DhairyaSethi) [inline src/libraries/types/DataTypes.sol]: let's make the caps uint40's such that `active` does not take a whole new slot

File: src/libraries/types/DataTypes.sol (hunk 3 above)
Comment (DhairyaSethi) [reply src/libraries/types/DataTypes.sol]: vis: we decided to inline instead

File: src/libraries/types/DataTypes.sol (hunk 5 above)
Comment (DhairyaSethi) [reply src/libraries/types/DataTypes.sol]: to go with uint40

File: tests/unit/Hub/Hub.Add.t.sol (hunk 8 above)
Comment (avniculae) [reply tests/unit/Hub/Hub.Add.t.sol:6]: apart from addAsset and add functions, I'm not sure we can test the safe casts bc operations will fail early in the validate functions [50e452a](https://github.com/aave/aave-v4/pull/591/commits/50e452a722c0b40c6e2257d95662f3ab7d29de0c)

File: src/libraries/types/DataTypes.sol (hunk 5 above)
Comment (DhairyaSethi) [reply src/libraries/types/DataTypes.sol]: caps can be packed more efficiently in spoke data struct, so we do it here as well

File: tests/unit/Hub/Hub.Add.t.sol (hunk 7 above)
Comment (avniculae) [reply tests/unit/Hub/Hub.Add.t.sol]: this is not a fuzz test, it's hardcoding the amounts

File: src/libraries/types/DataTypes.sol (hunk 14 above)
Comment (avniculae) [reply src/libraries/types/DataTypes.sol]: [bee414a](https://github.com/aave/aave-v4/pull/591/commits/bee414aa8875900f727dea092fce031c5c213f55)

File: src/libraries/types/DataTypes.sol (hunk 5 above)
Comment (avniculae) [reply src/libraries/types/DataTypes.sol]: [bee414a](https://github.com/aave/aave-v4/pull/591/commits/bee414aa8875900f727dea092fce031c5c213f55)

File: src/libraries/types/DataTypes.sol (hunk 4 above)
Comment (avniculae) [reply src/libraries/types/DataTypes.sol]: [7128866](https://github.com/aave/aave-v4/pull/591/commits/712886620db7e48a6ae9c94c7906201282ccbcf4)

File: tests/Base.t.sol (hunk 6 above)
Comment (avniculae) [reply tests/Base.t.sol]: [bee414a](https://github.com/aave/aave-v4/pull/591/commits/bee414aa8875900f727dea092fce031c5c213f55)

File: src/libraries/types/DataTypes.sol (hunk 13 above)
Comment (avniculae) [reply src/libraries/types/DataTypes.sol]: [7128866](https://github.com/aave/aave-v4/pull/591/commits/712886620db7e48a6ae9c94c7906201282ccbcf4)

File: src/libraries/types/DataTypes.sol (hunk 12 above)
Comment (avniculae) [reply src/libraries/types/DataTypes.sol]: [7128866](https://github.com/aave/aave-v4/pull/591/commits/712886620db7e48a6ae9c94c7906201282ccbcf4)

File: tests/gas/Spoke.Operations.gas.t.sol (hunk 11 above)
Comment (avniculae) [reply tests/gas/Spoke.Operations.gas.t.sol]: [7128866](https://github.com/aave/aave-v4/pull/591/commits/712886620db7e48a6ae9c94c7906201282ccbcf4)

File: src/libraries/logic/AssetLogic.sol (hunk 10 above)
Comment (avniculae) [reply src/libraries/logic/AssetLogic.sol]: [7128866](https://github.com/aave/aave-v4/pull/591/commits/712886620db7e48a6ae9c94c7906201282ccbcf4)

File: src/contracts/Hub.sol (hunk 9 above)
Comment (avniculae) [reply src/contracts/Hub.sol]: [7128866](https://github.com/aave/aave-v4/pull/591/commits/712886620db7e48a6ae9c94c7906201282ccbcf4)

File: tests/unit/HubConfigurator.t.sol [hunk 15]
```diff
@@ -5,6 +5,7 @@ import 'tests/unit/Hub/HubBase.t.sol';
 
 contract HubConfiguratorTest is HubBase {
   HubConfigurator public hubConfigurator;
+  using SafeCast for uint256;
```
Comment (DhairyaSethi) [inline tests/unit/HubConfigurator.t.sol]: nit: move up

Comment (avniculae) [reply tests/unit/HubConfigurator.t.sol]: removed it [ce6ddcf](https://github.com/aave/aave-v4/pull/591/commits/ce6ddcfa7856bd1641a1994a266a33adb0efc04d)
